### PR TITLE
VFB-101 pop up modals should have autofocus on relevant element and login should submit on enter

### PR DESCRIPTION
**Pull request by Jessica-Al**
## What's changed
It is possible to press enter to log in.
Each of the parcel action modals has an auto-selected option.
Additional delete parcel confirmation pop-up triggered when delete button pressed. Auto-selects confirm.
 
Change packing date: Date input is auto-selected, submit date change on enter
Change packing slot: Slot change input auto-selected dropdown closed (no submit on enter, opens dropdown)
Download day overview: Download button auto-selected
Download shopping list: Download button auto-selected
Download shipping labels: Quantity input auto-selected, submit changes on enter
Generate map: Generate map button auto-selected
Download driver overview: Driver name input auto-selected, download button activated on enter
Delete parcel: Delete parcel button auto-selected 

## Screenshots / Videos
New delete parcel confirmation
image

## Checklist
- [x] The ticket is up-to-date - Please document any deviations from the original approach if there is any.
- [x] I have documented the testing steps for QA
- [x] I have self-reviewed this PR
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [x] Make sure you've fixed all linting problems with `npm run lint_fix`
- [x] Make sure you've tested via `npm run test`

---
## AI generated change summary

The following is a summary of the changes in the PR generated by [What The Diff](https://whatthediff.ai/).
Delete the command below if you don't want this to be generataed.

wtd:summary

## Summary by CodeRabbit

- **New Features**
  - Enhanced authentication tests for improved accuracy.
  - Introduced new modal components for better organization and user interaction.
  - Added optional properties to various input components for improved usability.
  - New props for buttons to enhance form submission and focus behavior.

- **Bug Fixes**
  - Resolved issues related to button interaction in the login process.

- **Documentation**
  - Updated component props interfaces and comments for clarity.

- **Chores**
  - Streamlined code structure for better maintainability.

### CODE REVIEW

File: src/app/parcels/ActionBar/ActionModals/DateChangeModal.tsx [hunk 1]
```diff
@@ -8,18 +8,70 @@ import GeneralActionModal, {
 } from "./GeneralActionModal";
 import { Button } from "@mui/material";
 import SelectedParcelsOverview from "../SelectedParcelsOverview";
-import SingleDateInput, { DateInputProps } from "@/components/DateInputs/SingleDateInput";
+import SingleDateInput from "@/components/DateInputs/SingleDateInput";
 import dayjs, { Dayjs } from "dayjs";
 import { getStatusErrorMessageWithLogId } from "../Statuses";
 import { getDbDate } from "@/common/format";
 import { getUpdateErrorMessage, packingDateOrSlotUpdate } from "./CommonDateAndSlot";
+import { ParcelsTableRow } from "../../parcelsTable/types";
```
Comment (LeaHe188) [inline src/app/parcels/ActionBar/ActionModals/DateChangeModal.tsx]: import from @/app/...

File: src/app/parcels/ActionBar/ActionModals/DeleteParcelModal.tsx [hunk 2]
```diff
@@ -1,15 +1,85 @@
...
-import React, { useState } from "react";
-import GeneralActionModal, {
+import React, { useEffect, useRef, useState } from "react";
+import {
+    GeneralActionModal,
     ConfirmButtons,
     Heading,
     maxParcelsToShow,
     ActionModalProps,
 } from "./GeneralActionModal";
```
Comment (LeaHe188) [inline src/app/parcels/ActionBar/ActionModals/DeleteParcelModal.tsx:9]: unify the import format in all files:)

File: src/app/parcels/ActionBar/ActionModals/DeleteParcelModal.tsx [hunk 3]
```diff
@@ -1,15 +1,85 @@
...
+                >
+                    Delete
+                </Button>
+            </ConfirmButtons>
+            <Dialog
+                open={openConfirmationDialogue}
+                onClose={() => setOpenConfirmationDialogue(false)}
+            >
+                <DialogTitle id="alert-dialog-title">
+                    Please confirm you wish to delete the selected parcel/s.
```
Comment (LeaHe188) [inline src/app/parcels/ActionBar/ActionModals/DeleteParcelModal.tsx:68]: There's an extra slash:)

File: src/app/parcels/ActionBar/ActionModals/DriverOverviewModal.tsx [hunk 4]
```diff
@@ -69,6 +89,49 @@ const getPdfErrorMessage = (error: DriverOverviewError): string => {
     return `${errorMessage} LogId: ${error.logId}`;
 };
 
+const ModalContent: React.FC<ContentProps> = ({
```
Comment (LeaHe188) [inline src/app/parcels/ActionBar/ActionModals/DriverOverviewModal.tsx]: Do you want to specify the modal contents in different files? e.g DriverOverviewModalContent

File: src/app/parcels/ActionBar/ActionModals/GenerateMapModal.tsx [hunk 5]
```diff
@@ -1,9 +1,53 @@
...
 
+interface ContentProps {
+    uniquePostcodes: string[];
+    setErrorMessage: (message: string) => void;
+    setActionCompleted: (completed: boolean) => void;
+    mapsLinkForSelectedParcels: string;
+    setSuccessMessage: (message: string) => void;
+}
+
+const ModalContent: React.FC<ContentProps> = ({
```
Comment (LeaHe188) [inline src/app/parcels/ActionBar/ActionModals/GenerateMapModal.tsx]: same here, the modalcontent across different files are quite different yet sharing the same name

File: src/app/parcels/ActionBar/ActionModals/ShippingLabelModal.tsx [hunk 6]
```diff
@@ -15,23 +15,39 @@ import { getStatusErrorMessageWithLogId } from "../Statuses";
...
+    onPdfCreationCompleted: () => void;
+    onPdfCreationFailed: (pdfError: ShippingLabelError) => void;
+    onLabelQuantityChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
+    duplicateDownloadedPostcodes: (string | null)[];
+}
+
+const ShippingLabelsInput = React.forwardRef<HTMLInputElement, ShippingLabelsInputProps>(
+    (props, elementToFocusRef) => {
+        return (
+            <>
```
Comment (LeaHe188) [inline src/app/parcels/ActionBar/ActionModals/ShippingLabelModal.tsx:37]: Formatting problem (a minor one): do you want to leave some gap between each component? They look a bit packed

File: src/app/parcels/ActionBar/ActionModals/SlotChangeModal.tsx [hunk 7]
```diff
@@ -31,21 +42,57 @@ const SlotChangeInput: React.FC<SlotInputProps> = ({ packingSlotsLabelsAndValues
...
     const [actionCompleted, setActionCompleted] = useState(false);
     const [errorMessage, setErrorMessage] = useState<string | null>(null);
     const [successMessage, setSuccessMessage] = useState<string | null>(null);
 
     const [packingSlots, setPackingSlots] = useState<[string, string][]>([]);
-    const [slot, setSlot] = useState<string>("");
     const [displayModal, setDisplayModal] = useState<boolean>(true);
     const [warningMessage, setWarningMessage] = useState<string>("");
 
+    let slot = "";
```
Comment (LeaHe188) [inline src/app/parcels/ActionBar/ActionModals/SlotChangeModal.tsx]: are these five lines ever used? If not then it's a better idea to delete them

File: src/app/parcels/ActionBar/ActionModals/SlotChangeModal.tsx [hunk 8]
```diff
@@ -108,39 +155,19 @@ const SlotChangeModal: React.FC<ActionModalProps> = (props) => {
...
-                        <Button variant="contained" onClick={onSlotSubmit}>
-                            Change
-                        </Button>
-                    </ConfirmButtons>
-                ) : (
-                    <></>
-                )
-            }
-            contentAboveButton={
-                displayModal ? (
```
Comment (LeaHe188) [inline src/app/parcels/ActionBar/ActionModals/SlotChangeModal.tsx:127]: displayModal && (...)

and wrap the parent div outside

File: src/components/DataInput/DropdownListInput.tsx [hunk 9]
```diff
@@ -9,16 +9,24 @@ interface Props<Value> {
     defaultValue?: Value;
     onChange?: (event: SelectChangeEvent<Value>) => void;
     selectLabelId: string;
+    focusOnDropdown?: boolean;
 }
 
 const DropdownListInput = <Value,>(props: Props<Value>): React.ReactElement => {
```
Comment (LeaHe188) [inline src/components/DataInput/DropdownListInput.tsx:15]: extra comma:)

File: src/components/DataInput/DropdownListInput.tsx [hunk 10]
```diff
@@ -9,16 +9,24 @@ interface Props<Value> {
...
+    const elementToFocusRef = React.useRef<HTMLInputElement>(null);
```
Comment (LeaHe188) [inline src/components/DataInput/DropdownListInput.tsx]: DropDownListInput is actually used in many files, and so it's a good idea to make sure all the files still survive with those changes

File: src/components/DateInputs/SingleDateInput.tsx [hunk 11]
```diff
@@ -10,22 +10,27 @@ export interface DateInputProps {
     setDate: (date: Dayjs) => void;
 }
 
-const SingleDateInput: React.FC<DateInputProps> = (dateInputProps) => {
-    const setSingleDate = (date: Dayjs | null): void => {
-        if (date) {
-            dateInputProps.setDate(date);
-        }
-    };
+const SingleDateInput = React.forwardRef<HTMLInputElement, DateInputProps>(
```
Comment (LeaHe188) [inline src/components/DateInputs/SingleDateInput.tsx:13]: Same here, one more argument could trigger error else where, so make sure this is tested

File: src/components/PdfButton/PdfButton.tsx [hunk 12]
```diff
@@ -55,12 +60,23 @@ const PdfButton = <Data, ErrorType extends string>({
         saveAs(blob, formatName ? formatFileName(data.fileName) : data.fileName);
         onPdfCreationCompleted();
     };
+
+    const elementToFocusRef = useRef<HTMLButtonElement>(null);
```
Comment (LeaHe188) [inline src/components/PdfButton/PdfButton.tsx]: Same, you may want to test all files that are involved

File: src/components/DataInput/FreeFormTextInput.tsx [hunk 13]
```diff
@@ -16,9 +16,12 @@ interface Props {
...
     className?: string;
     fullWidth?: boolean;
+    margin?: "dense" | "normal" | "none";
 }
 
-const FreeFormTextInput: React.FC<Props> = (props) => {
-    return <TextField {...props} />;
-};
+const FreeFormTextInput = React.forwardRef<HTMLInputElement, Props>((props, focusRef) => {
+    return <TextField {...props} inputRef={focusRef} />;
```
Comment (LeaHe188) [inline src/components/DataInput/FreeFormTextInput.tsx:23]: same here

File: src/app/parcels/ActionBar/ActionModals.test.tsx [hunk 14]
```diff
@@ -58,14 +54,16 @@ describe("Actions", () => {
...
                 successMessage={null}
-                actionButton={mockActionButton}
-                contentAboveButton={mockContentAboveButton}
                 header="TestModal"
                 isOpen={false}
                 headerId="test-modal-header"
-            />
+            >
+                <p> Test Content </p>
+                <button data-testid="action-button" onClick={mockOnSubmit}>
```
Comment (LeaHe188) [inline src/app/parcels/ActionBar/ActionModals.test.tsx:63]: Some files add hashtags in front of the testid, i.e. #action-button, maybe it's a good idea to unify these to match the expression

**Review by LeaHe188 — COMMENTED** [review]
LGTM! I've left some comments above:)

File: src/app/parcels/ActionBar/ActionModals/DateChangeModal.tsx [hunk 15]
```diff
@@ -13,13 +13,61 @@ import dayjs, { Dayjs } from "dayjs";
...
+    setDate,
+    selectedParcels,
+    maxParcelsToShow,
+    warningMessage,
+}) => {
+    const elementToFocusRef = useRef<HTMLInputElement>(null);
+
+    useEffect(() => {
+        elementToFocusRef.current?.focus();
+    }, []);
```
Comment (Stuart-Knowles-SW) [inline src/app/parcels/ActionBar/ActionModals/DateChangeModal.tsx]: Can we name this for what it is, i.e., `dateChangeInputRef`? It's pretty clear it's used to focus the element, would be nice to know which element at a glance. This is both here and elsewhere :)

File: src/app/parcels/ActionBar/ActionModals/SlotChangeModal.tsx [hunk 16]
```diff
@@ -31,11 +42,43 @@ const SlotChangeInput: React.FC<SlotInputProps> = ({ packingSlotsLabelsAndValues
...
+    displayModal,
+    packingSlots,
+    setSlot,
+    selectedParcels,
+    warningMessage,
+    onClose,
+    onSlotSubmit,
+}) => {
+    return (
+        displayModal && (
```
Comment (Stuart-Knowles-SW) [inline src/app/parcels/ActionBar/ActionModals/SlotChangeModal.tsx]: can we move the boolean check for `displayModal` outside the content and include it with the `actionComplete` check?

**Review by Stuart-Knowles-SW — COMMENTED** [review]
Really great work! Just a couple of small things :)

**Review by Stuart-Knowles-SW — APPROVED** [review]
LGTM!

**Pull request merged**